File: content/develop/use-cases/prefetch-cache/dotnet/MockPrimaryStore.cs
using System.Collections.Concurrent;

namespace PrefetchCacheDemo;

/// <summary>
/// Mock primary data store for the prefetch-cache demo.
///
/// This stands in for a source-of-truth database (Postgres, MySQL,
/// Mongo, etc.) that holds reference data the application serves to
/// users.
///
/// Every mutation appends a change event to an in-process queue, which
/// the sync worker drains and applies to Redis. In a real system the
/// queue is replaced by a CDC pipeline — Redis Data Integration,
/// Debezium plus a lightweight consumer, or an equivalent tool that
/// tails the source's binlog/WAL and pushes changes into Redis.
///
/// The store also exposes <see cref="ReadLatencyMs"/> so the demo can
/// illustrate how much slower a direct primary read would be than a
/// Redis hit.
/// </summary>
public class MockPrimaryStore
{
    public int ReadLatencyMs { get; }

    private readonly object _lock = new();
    private long _reads;
    private readonly Dictionary<string, Dictionary<string, string>> _records;
    private readonly BlockingCollection<ChangeEvent> _changes = new(new ConcurrentQueue<ChangeEvent>());

    public MockPrimaryStore(int readLatencyMs = 80)
    {
        ReadLatencyMs = readLatencyMs;
        _records = new Dictionary<string, Dictionary<string, string>>(StringComparer.Ordinal)
        {
            ["cat-001"] = new()
            {
                ["id"] = "cat-001",
                ["name"] = "Beverages",
                ["display_order"] = "1",
                ["featured"] = "true",
                ["parent_id"] = "",
            },
            ["cat-002"] = new()
            {
                ["id"] = "cat-002",
                ["name"] = "Bakery",
                ["display_order"] = "2",
                ["featured"] = "true",
                ["parent_id"] = "",
            },
            ["cat-003"] = new()
            {
                ["id"] = "cat-003",
                ["name"] = "Pantry Staples",
                ["display_order"] = "3",
                ["featured"] = "false",
                ["parent_id"] = "",
            },
            ["cat-004"] = new()
            {
                ["id"] = "cat-004",
                ["name"] = "Frozen",
                ["display_order"] = "4",
                ["featured"] = "false",
                ["parent_id"] = "",
            },
            ["cat-005"] = new()
            {
                ["id"] = "cat-005",
                ["name"] = "Specialty Cheeses",
                ["display_order"] = "5",
                ["featured"] = "false",
                ["parent_id"] = "cat-002",
            },
        };
    }

    public List<string> ListIds()
    {
        lock (_lock)
        {
            var ids = _records.Keys.ToList();
            ids.Sort(StringComparer.Ordinal);
            return ids;
        }
    }

    /// <summary>Return every record. Used by the cache's bulk-load path on startup.</summary>
    public List<Dictionary<string, string>> ListRecords()
    {
        Thread.Sleep(ReadLatencyMs);
        lock (_lock)
        {
            Interlocked.Increment(ref _reads);
            return _records.Values
                .Select(r => new Dictionary<string, string>(r, StringComparer.Ordinal))
                .ToList();
        }
    }

    /// <summary>Single-record read. Not on the demo's normal read path.</summary>
    public Dictionary<string, string>? Read(string entityId)
    {
        Thread.Sleep(ReadLatencyMs);
        lock (_lock)
        {
            Interlocked.Increment(ref _reads);
            return _records.TryGetValue(entityId, out var record)
                ? new Dictionary<string, string>(record, StringComparer.Ordinal)
                : null;
        }
    }

    public bool AddRecord(Dictionary<string, string> record)
    {
        if (!record.TryGetValue("id", out var entityId) || string.IsNullOrEmpty(entityId?.Trim()))
        {
            return false;
        }
        entityId = entityId.Trim();
        lock (_lock)
        {
            if (_records.ContainsKey(entityId))
            {
                return false;
            }
            _records[entityId] = new Dictionary<string, string>(record, StringComparer.Ordinal);
            // Emit while the lock is held so the queue order matches the
            // mutation order. Two concurrent callers cannot interleave
            // mutation A -> mutation B -> emit B -> emit A.
            EmitChangeLocked(ChangeOp.Upsert, entityId, new Dictionary<string, string>(record, StringComparer.Ordinal));
        }
        return true;
    }

    public bool UpdateField(string entityId, string field, string value)
    {
        lock (_lock)
        {
            if (!_records.TryGetValue(entityId, out var record))
            {
                return false;
            }
            record[field] = value;
            EmitChangeLocked(
                ChangeOp.Upsert,
                entityId,
                new Dictionary<string, string>(record, StringComparer.Ordinal));
        }
        return true;
    }

    public bool DeleteRecord(string entityId)
    {
        lock (_lock)
        {
            if (!_records.Remove(entityId))
            {
                return false;
            }
            EmitChangeLocked(ChangeOp.Delete, entityId, null);
        }
        return true;
    }

    /// <summary>Block up to <paramref name="timeout"/> for the next change event.</summary>
    public ChangeEvent? NextChange(TimeSpan timeout)
    {
        if (_changes.TryTake(out var change, timeout))
        {
            return change;
        }
        return null;
    }

    public long Reads => Interlocked.Read(ref _reads);

    public void ResetReads() => Interlocked.Exchange(ref _reads, 0);

    /// <summary>
    /// Append a change event to the feed. Caller must hold <c>_lock</c>.
    ///
    /// <see cref="BlockingCollection{T}.Add(T)"/> is itself thread-safe
    /// and never tries to acquire <c>_lock</c>, so calling it while
    /// holding the records lock cannot deadlock. Holding the lock here
    /// is what guarantees that the queue order matches the order in
    /// which the records dict was mutated.
    /// </summary>
    private void EmitChangeLocked(ChangeOp op, string entityId, Dictionary<string, string>? fields)
    {
        // Use millisecond-precision unix timestamp so the sync-lag
        // metric is in the same shape as the Python reference.
        var timestampMs = (double) DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        _changes.Add(new ChangeEvent(op, entityId, fields, timestampMs));
    }
}

File: content/develop/use-cases/prefetch-cache/java-lettuce/MockPrimaryStore.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Mock primary data store for the prefetch-cache demo.
 *
 * <p>This stands in for a source-of-truth database (Postgres, MySQL,
 * Mongo, etc.) that holds reference data the application serves to
 * users.</p>
 *
 * <p>Every mutation appends a change event to an in-process queue,
 * which the sync worker drains and applies to Redis. In a real system
 * the queue is replaced by a CDC pipeline — Redis Data Integration,
 * Debezium plus a lightweight consumer, or an equivalent tool that
 * tails the source's binlog/WAL and pushes changes into Redis.</p>
 *
 * <p>The store also exposes a read-latency knob so the demo can
 * illustrate how much slower a direct primary read would be than a
 * Redis hit.</p>
 */
public class MockPrimaryStore {

    public static final String CHANGE_OP_UPSERT = "upsert";
    public static final String CHANGE_OP_DELETE = "delete";

    private final int readLatencyMs;
    private final AtomicLong reads = new AtomicLong();
    private final Object lock = new Object();
    private final Map<String, Map<String, String>> records = new LinkedHashMap<>();
    private final LinkedBlockingQueue<Map<String, Object>> changes = new LinkedBlockingQueue<>();

    public MockPrimaryStore() {
        this(80);
    }

    public MockPrimaryStore(int readLatencyMs) {
        this.readLatencyMs = readLatencyMs;
        seed("cat-001", "Beverages", "1", "true", "");
        seed("cat-002", "Bakery", "2", "true", "");
        seed("cat-003", "Pantry Staples", "3", "false", "");
        seed("cat-004", "Frozen", "4", "false", "");
        seed("cat-005", "Specialty Cheeses", "5", "false", "cat-002");
    }

    public int getReadLatencyMs() {
        return readLatencyMs;
    }

    /** Sorted list of every record id. No sleep, no counter increment (metadata-only). */
    public List<String> listIds() {
        List<String> ids;
        synchronized (lock) {
            ids = new ArrayList<>(records.keySet());
        }
        Collections.sort(ids);
        return ids;
    }

    /** Return every record. Used by the cache's bulk-load path on startup. */
    public List<Map<String, String>> listRecords() {
        sleepReadLatency();
        synchronized (lock) {
            reads.incrementAndGet();
            List<Map<String, String>> out = new ArrayList<>(records.size());
            for (Map<String, String> record : records.values()) {
                out.add(new LinkedHashMap<>(record));
            }
            return out;
        }
    }

    /** Single-record read. Not on the demo's normal read path. */
    public Map<String, String> read(String entityId) {
        sleepReadLatency();
        synchronized (lock) {
            reads.incrementAndGet();
            Map<String, String> record = records.get(entityId);
            return record == null ? null : new LinkedHashMap<>(record);
        }
    }

    public boolean addRecord(Map<String, String> record) {
        if (record == null) return false;
        String entityId = record.getOrDefault("id", "").trim();
        if (entityId.isEmpty()) return false;
        synchronized (lock) {
            if (records.containsKey(entityId)) {
                return false;
            }
            records.put(entityId, new LinkedHashMap<>(record));
            // Emit while the lock is held so the queue order matches
            // the mutation order. Two concurrent callers cannot
            // interleave mutation A → mutation B → emit B → emit A.
            emitChangeLocked(CHANGE_OP_UPSERT, entityId, new LinkedHashMap<>(record));
        }
        return true;
    }

    public boolean updateField(String entityId, String field, String value) {
        synchronized (lock) {
            Map<String, String> record = records.get(entityId);
            if (record == null) {
                return false;
            }
            record.put(field, value);
            emitChangeLocked(CHANGE_OP_UPSERT, entityId, new LinkedHashMap<>(record));
        }
        return true;
    }

    public boolean deleteRecord(String entityId) {
        synchronized (lock) {
            if (!records.containsKey(entityId)) {
                return false;
            }
            records.remove(entityId);
            emitChangeLocked(CHANGE_OP_DELETE, entityId, null);
        }
        return true;
    }

    /**
     * Block up to {@code timeoutMs} for the next change event. Returns
     * {@code null} on timeout.
     */
    public Map<String, Object> nextChange(long timeoutMs) {
        try {
            return changes.poll(timeoutMs, TimeUnit.MILLISECONDS);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            return null;
        }
    }

    public long reads() {
        return reads.get();
    }

    public void resetReads() {
        reads.set(0);
    }

    /**
     * Append a change event to the feed. Caller must hold {@link #lock}.
     *
     * <p>{@link LinkedBlockingQueue#offer(Object)} is itself thread-safe
     * and never tries to acquire {@link #lock}, so calling it while
     * holding the records lock cannot deadlock. Holding the lock here
     * is what guarantees the queue order matches the order in which
     * the records map was mutated.</p>
     */
    private void emitChangeLocked(String op, String entityId, Map<String, String> fields) {
        Map<String, Object> event = new HashMap<>();
        event.put("op", op);
        event.put("id", entityId);
        event.put("fields", fields);
        event.put("timestamp_ms", (double) System.currentTimeMillis());
        changes.offer(event);
    }

    private void sleepReadLatency() {
        if (readLatencyMs <= 0) return;
        try {
            Thread.sleep(readLatencyMs);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    private void seed(String id, String name, String displayOrder, String featured, String parentId) {
        Map<String, String> record = new LinkedHashMap<>();
        record.put("id", id);
        record.put("name", name);
        record.put("display_order", displayOrder);
        record.put("featured", featured);
        record.put("parent_id", parentId);
        records.put(id, record);
    }
}
